httpsync/message.go

```diff
@@ -0,0 +1,124 @@
+package httpsync
+
+import (
+	"bytes"
+	"errors"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/codec/dagjson"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+	ic "github.com/libp2p/go-libp2p-core/crypto"
+)
+
+var typeSystem *schema.TypeSystem = createTypeSystem()
+
+func createTypeSystem() *schema.TypeSystem {
+	ts := schema.TypeSystem{}
+	ts.Init()
+	ts.Accumulate(schema.SpawnBytes("Bytes"))
+	ts.Accumulate(schema.SpawnLink("Link"))
+	ts.Accumulate(schema.SpawnStruct("SignedHead",
+		[]schema.StructField{
+			schema.SpawnStructField("head", "Link", false, false),
+			schema.SpawnStructField("sig", "Bytes", false, false),
+			schema.SpawnStructField("pubkey", "Bytes", false, false),
+		},
+		schema.SpawnStructRepresentationMap(nil),
+	))
+
+	return &ts
+}
+
+func SignedHeadSchema() schema.Type {
+	return typeSystem.TypeByName("SignedHead")
+}
+
+// signedHead is the signed envelope of the head CID. It includes the
+// public key of the signer so the receiver can verify it and convert it to a
+// peer id. Note, the receiver is not required to use the provided public key.
+type signedHead struct {
+	Head   cidlink.Link
+	Sig    []byte
+	Pubkey []byte
+}
+
+// NewSignedEnvelope returns a new encoded SignedHead
+func NewEncodedSignedHead(cid cid.Cid, privKey ic.PrivKey) ([]byte, error) {
+	sig, err := privKey.Sign(cid.Bytes())
+	if err != nil {
+		return nil, err
+	}
+
+	pubKeyBytes, err := ic.MarshalPublicKey(privKey.GetPublic())
+	if err != nil {
+		return nil, err
+	}
+
+	envelop := &signedHead{
+		Head:   cidlink.Link{Cid: cid},
+		Sig:    sig,
+		Pubkey: pubKeyBytes,
+	}
+	node := bindnode.Wrap(envelop, SignedHeadSchema())
+	var buf bytes.Buffer
+	err = dagjson.Encode(node.Representation(), &buf)
+	if err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// OpenSignedHead returns the cid from the encoded signed head given a public key
+func OpenSignedHead(pubKey ic.PubKey, SignedHead io.Reader) (cid.Cid, error) {
```

Comment:
If this function is not used outside the `httpsync` package then please do not export it.